rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+  - Otherwise, the body of `with attrs;` must start on a new line without any additional indentation.
+- If the body is `{ ... }`, `[ ... ]`, `( ... )` or `'' ... ''`
+  and `with attrs; …` is parenthesised or to the right side of a binding (e.g. `(with; [ …` or `foo = with; [ …`),
+  then `with` must start on the same line.
+  - Otherwise it must start on a new line.
+
+**Examples**
+
+```nix
+{
+  # Good
+  foo = with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Good
+  foo =
+    with foo;
+    with bar; [
+      # multiline
+      baz
+    ];
+  
+  # Good
+  foo =
+    with bar;
+    baz foo {
+      # multiline
+      qux = 10;
+    };
+
+  # Good
+  foo =
+    with bar;
+    if cond then
+      foo
+    else
+      bar;
+  
+  # Bad
+  foo = assert qux; with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Bad
+  foo = with bar;
+    [
+      # multiline
+      baz
+    ];
+
+  # Bad
+  foo =
+    with bar; [
+      # multiline
+      baz
+    ];
+    
+  # Good
+  [
+    qux
+    quux
+  ]
+  ++ (with pkgs; [
+    baz
+    blorp
+  ]);
+}
+```
+
+### let-in
+
+Let bindings must always have this form:
+```
+let
+  <name1> = <value1>;
+  <name2> = <value2>;
+  ...
+in
+<body>
+```
+
+- Let bindings are *always* multiline.
+- Each binding is indented and starts on its own line.
+  For more details, see the [bindings section](#bindings).
+- The `<body>` always starts on a new line and is not indented.
+
+**Examples**
+
+```nix
+let
+  foo = "bar";
+in
+func foo;
+
+let
+  foo = "bar";
+in
+{
+  inherit foo;
+  baz = "smth";
+}
+
+let
+  foo = "bar";
+in
+if foo == "bar" then
+  "hello"
+else
+  "world"
+```
+
+**Alternatives**
+
+- To allow having the `<body>` be on the same line as the `in`:
+  ```
+  let
+    <name1> = <value1>;
+    <name2> = <value2>;
+    ...
+  in <body>
+  ```
+
+  In particular when `<body>` is an identifier, list, attribute set and/or others.
+
+  Problems with this alternative:
+  - It leads to larger diffs when inserting something after the `in`
+  - The formatting can change when `<body>` is updated
+  - It's less consistent, since the formatting depends on the `<body>`
+
+- The body could be indented by a level
+  ```
+  let
+    <name1> = <value1>;
+    <name2> = <value2>;
+    ...
+  in
+    <body>
+  ```
+  
+  Problems with this alternative:
+  - Leads to indentation creeps
+  - Inconsistent with other expressions that have a `<body>` that is "returned"
+  - Favors a style where the body starts on the same line as the in for some values (e.g. attribute sets) to reduce an indentation level, see above.
+
+### Attribute sets and lists
+
+- Brackets and braces must always have a space (or line break) on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - Empty lists and attribute sets are written as `[ ]` and `{ }`, respectively.
+- Lists and attribute sets with multiple items should be liberally expanded.
+  - They can only be on a single line if they fit on the line and contain few enough elements.
```

Comment:
This seems to be a rule that would have to take the _input_ formatting into account. We're trying to really minimise such rules since they can lead to differing opinions. However we do already allow empty newlines from the input format to be persisted. And what you're suggesting feels like another good case where we can make an exception and allow it.

We'll give this a try and follow up later!

Reply:
After sleeping over the idea, an alternative to tracking the input would be to implement formatting directives. They have been discussed before and are left open as an option in the RFC. I'd honestly prefer that over adding another "impurity" to the parser.

Please provide some real-world examples of situations where you'd want the opposite of what Nixfmt does w.r.t. expansion, to better understand the tradeoffs between each of the approaches.

Reply:
The most critical IMO is that it forces this declaration:

```Nix
    custom.packages.allowedUnfree = [
      "spotify"
    ];
```

into

```Nix
    custom.packages.allowedUnfree = [ "spotify" ];
```

while the list being multi-line was intentional to allow adding additional elements without diff noise.

Reply:
Is this still preserved, and if so is it good enough for that case?
```nix
custom.packages.allowedUnfree = [
  "spotify"
  # ...
];
```

Reply:
Yes, as a work-around, adding a comment would force the multi-line layout. An empty line should work too btw (FWIW):

```nix
custom.packages.allowedUnfree = [

  "spotify"
];
```

While I make plenty of use of comments to force multiline layouts in the test suite, generally this isn't that great and we should try to find a better solution. As for the comment directives proposal, without giving too much thought to syntax for now it could look like this:

```nix
# @nixfmt:expand
custom.packages.allowedUnfree = [
  "spotify"
];
```

-----

However, I'd argue that a single-line `custom.packages.allowedUnfree = [ spotify ];` is not *that* bad. Having a single item with the expectation to add more in the future is pretty rare IME. Most of the time it stays at one item, or it already is more than one in the first place. So even when you run into the case where you only have one item and expect to add more in the future, as soon as the second item is added it will use the expanded layout anyways.

Reply:
> Having a single item with the expectation to add more in the future is pretty rare IME

I would like to offer my own (anecdotal) experience as counter-point. People who learn nix/are not that experienced with it (a lot of whom use flakes for devshells for example) do this in my (again anecdotal) experience.

I myself do the same with my own nix configs, for purposes just like allowedUnfreePackages.

Reply:
Advantages of an inline comment over a comment directive, in this particular case:
* No special syntax to remember
* Comment is literally a placeholder for more things, in exactly the place that the new things would go
* If `...` isn't evocative enough for you, you can make it say literally anything you want

Disadvantages:
* Less consistent with other cases where a comment directive might be more appropriate?

Reply:
We discussed this today and agreed that it would be okay to preserve such multiline single-element lists. There's nothing in the RFC that disallows this, and we added a line that indicates that it would be allowed: https://github.com/NixOS/rfcs/pull/166/commits/463a858e57e4d68f4bfb1d38f20d8ac852ac4574